### PR TITLE
♻️ Transform function outputs dict of DataFrames

### LINKED ISSUE #320: Transform func should output a dict of dataframes
The transform function currently outputs a dataframe that is the result of merging all of the extracted dataframes. There are cases where it doesn't make sense to merge all of the data together into a single df (e.g. family relationships)

The transform function should output a dict of dataframes which follows this format:

```
{
    'family_relationship': family_rel_df,
    '<another target concept>': filtered_df_for_some_concept,
    'default': single_merged_df_with_everything_else
}
```
Let's say transform stage is building `family_relationship` target concept instances. It will look at the transform func output dict. If there is a key in the dict matching the name of the target concept, `family_relationship`, then it will use that key's value or the dataframe containing the `family_relationships` to build the target concept instances. Otherwise, it will look for the `default` (or whatever we name it) key and use the DataFrame contained in that key's value.

### PR DESCRIPTION
**Pull request by znatty22**
Closes #320

### CODE REVIEW

File: templates/my_study/transform_module.py [hunk 1]
```diff
@@ -13,9 +13,44 @@
...
+
+    If not possible to merge all DataFrames into a single DataFrame then
+    you can return a dict that looks something like this:
+
+    {
+        '<name of target concept>': df_for_<target_concept>,
+        'default': all_merged_data_df
+    }
+
+    Target concept instances will be built from the default DataFrame, unless
```
Comment (fiendish) [inline templates/my_study/transform_module.py]:
```suggestion
    Target concept instances will be built from the default DataFrame unless
```

File: kf_lib_data_ingest/etl/transform/guided.py [hunk 2]
```diff
@@ -26,6 +34,9 @@ class GuidedTransformStage(TransformStage):
     def __init__(self, transform_function_path, *args, **kwargs):
         self.transform_module = TransformModule(transform_function_path)
         super().__init__(*args, **kwargs)
+        self.transform_func_dir = os.path.join(self.stage_cache_dir,
+                                               'transform_function_output')
+        os.makedirs(self.transform_func_dir, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (znatty22) [reply kf_lib_data_ingest/etl/transform/guided.py]: Will move it 👍

File: kf_lib_data_ingest/etl/transform/guided.py [hunk 3]
```diff
@@ -66,66 +78,119 @@ def _apply_transform_funct(self, data_dict):
...
 
         For example,
 
         This:
 
-        |CONCEPT.PARTICIPANT.UNIQUE_KEY | CONCEPT.PARTICIPANT.GENDER| CONCEPT.BIOSPECIMEN.UNIQUE_KEY| # noqa E501
-        |-------------------------------|---------------------------|-------------------------------| # noqa E501
-        |        P1                     |            Female         |            B1                 | # noqa E501
-        |        P2                     |            Male           |            B2                 | # noqa E501
+        {
```
Comment (znatty22) [inline kf_lib_data_ingest/etl/transform/guided.py]: Updated docstring to be more realistic

File: kf_lib_data_ingest/etl/transform/guided.py [hunk 4]
```diff
@@ -66,66 +78,119 @@ def _apply_transform_funct(self, data_dict):
...
                 {
-                    'id': 'P1',
+                    'id': 'P1-CDH-400',
                     'properties': {
-                        'gender': 'Female',
+                         'age_at_event_days': 400,
+                         'source_text': 'CDH',
                          ...
+                    },
+                    links: {
```
Comment (znatty22) [inline kf_lib_data_ingest/etl/transform/guided.py]: Links format will change in #319, docstring will be updated in that PR

File: kf_lib_data_ingest/etl/transform/guided.py [hunk 5]
```diff
@@ -66,66 +78,119 @@ def _apply_transform_funct(self, data_dict):
...
+                        'participant_id': 'P1'
                     }
                 },
                 {
-                    'id': 'P2',
+                    'id': 'P2-CDH-500',
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/guided.py]:
```suggestion
                    'id': 'P2-CNS-500',
```

File: templates/my_study/transform_module.py [hunk 6]
```diff
@@ -13,9 +13,44 @@
...
 def transform_function(mapped_df_dict):
-    # Expected to return a single merged dataframe
+    """
+    Merge DataFrames in mapped_df_dict into 1 DataFrame if possible.
 
-    return list(mapped_df_dict.values())[0]
+    Return a dict that looks like this:
+
+    {
+        'default': all_merged_data_df
```
Comment (fiendish) [inline templates/my_study/transform_module.py]: DEFAULT_KEY?

File: templates/my_study/transform_module.py [hunk 7]
```diff
@@ -13,9 +13,44 @@
...
+    another DataFrame is explicitly provided via a key, value pair in the
+    output dict. They key must match the name of an existing target concept.
+    The value will be the DataFrame to use when building instances of the
+    target concept.
+
+    A typical example would be:
+
+    {
+        'family_relationship': family_relationship_df,
+        'default': all_merged_data_df
```
Comment (fiendish) [inline templates/my_study/transform_module.py]: lol

**Pull request merged**